### PR TITLE
feat: move ocr detection into preprocessor

### LINKED ISSUE #672: fix: optimise latency of OCR detector
Optimising latency of OCR detector by comparing: [pytesseract](https://cloud.jina.ai/executor/ypt3z5tu?random=4067) (CPU), [keras OCR](TestExecutorKerasOCR) (CPU & GPU), [PaddlePaddleOCR](https://cloud.jina.ai/executor/rt321xqd?random=1970) (CPU & GPU), [Paddlepaddle OCR other](https://cloud.jina.ai/executor/78yp7etm) (CPU & GPU). All, except for last one are built based on OCR detector of now.

Recommendation:
- use [PaddlePaddleOCR](https://cloud.jina.ai/executor/rt321xqd?random=1970) executor on CPU as an external executor with auto-scaling
- test quality of paddle paddle vs keras OCR

Dashboard from WOLF:
 - pytesseract on CPU: ~150 ms
 - keras OCR on GPU: ~150 ms
 - keras OCR on CPU: ~2 seconds
 - PaddlePaddleOCR on CPU: ~300 ms
 - PaddlePaddleOCR on GPU: ~20 mins
 - PaddlepaddleOCR other on CPU: ~600ms
 - PaddlepaddleOCR other on GPU: ~400ms
Screenshot 2022-11-16 at 10 15 25
Screenshot 2022-11-16 at 10 17 42
Screenshot 2022-11-16 at 11 18 03

Benchmarking latency by calling Jina client:

```
Latency of tesseract for 10 docs: 2.1132253408432007s
QPS of tesseract for 10 docs: 0.6241443915326548s
P0 of tesseract for 10 docs: 4.093590021133423
P50 of tesseract for 10 docs: 7.898972988128662
P75 of tesseract for 10 docs: 8.157279014587402
P85 of tesseract for 10 docs: 8.346951246261597
P90 of tesseract for 10 docs: 8.533043146133423
P95 of tesseract for 10 docs: 8.822581052780151
P99 of tesseract for 10 docs: 12.86188817024231
P99.9 of tesseract for 10 docs: 12.86188817024231


Latency of keras_cpu for 10 docs: 21.374321899414063s


Latency of keras_gpu for 10 docs: 2.25632716178894s
QPS of keras_gpu for 10 docs: 0.6055373692558534s
P0 of keras_gpu for 10 docs: 4.872711181640625
P50 of keras_gpu for 10 docs: 7.570810079574585
P75 of keras_gpu for 10 docs: 10.032758951187134
P85 of keras_gpu for 10 docs: 10.0690279006958
P90 of keras_gpu for 10 docs: 10.142452239990234
P95 of keras_gpu for 10 docs: 11.292683124542236
P99 of keras_gpu for 10 docs: 11.550961971282959
P99.9 of keras_gpu for 10 docs: 11.550961971282959


Latency of paddle_cpu for 10 docs: 3.5214644145965575s
QPS of paddle_cpu for 10 docs: 0.33373029454304487s
P0 of paddle_cpu for 10 docs: 8.5104079246521
P50 of paddle_cpu for 10 docs: 14.329988241195679
P75 of paddle_cpu for 10 docs: 14.416431188583374
P85 of paddle_cpu for 10 docs: 14.472574949264526
P90 of paddle_cpu for 10 docs: 14.66151237487793
P95 of paddle_cpu for 10 docs: 19.2405047416687
P99 of paddle_cpu for 10 docs: 23.46496891975403
P99.9 of paddle_cpu for 10 docs: 23.46496891975403


paddle paddle GPU wouldn't stop


Latency of paddle_cpu_other for 10 docs: 6.441506142616272s
QPS of paddle_cpu_other for 10 docs: 0.16484086077726945s
P0 of paddle_cpu_other for 10 docs: 20.983601093292236
P50 of paddle_cpu_other for 10 docs: 30.057352781295776
P75 of paddle_cpu_other for 10 docs: 30.311936140060425
P85 of paddle_cpu_other for 10 docs: 31.1715190410614
P90 of paddle_cpu_other for 10 docs: 31.965965032577515
P95 of paddle_cpu_other for 10 docs: 35.08013677597046
P99 of paddle_cpu_other for 10 docs: 36.348729848861694
P99.9 of paddle_cpu_other for 10 docs: 36.348729848861694

Latency of paddle_gpu_other for 10 docs: 4.652856059074402s
QPS of paddle_gpu_other for 10 docs: 0.23725539241543467s
P0 of paddle_gpu_other for 10 docs: 12.263688802719116
P50 of paddle_gpu_other for 10 docs: 20.91442036628723
P75 of paddle_gpu_other for 10 docs: 21.25279211997986
P85 of paddle_gpu_other for 10 docs: 21.553786993026733
P90 of paddle_gpu_other for 10 docs: 21.945390224456787
P95 of paddle_gpu_other for 10 docs: 25.055743932724
P99 of paddle_gpu_other for 10 docs: 26.566694974899292
P99.9 of paddle_gpu_other for 10 docs: 26.566694974899292
```

### PR DESCRIPTION
**Pull request by joschkabraun**
Closes #672 by moving OCR detection into preprocessor. It will be applied only during indexing for image documents. It also uses PaddlePaddleOCR.

### CODE REVIEW

File: now/app/text_to_video/flow-video-clip-high-performance.yml [hunk 1]
```diff
@@ -41,24 +40,8 @@ executors:
...
-    external: ${{ ENV.IS_REMOTE_DEPLOYMENT }}
-    needs: preprocessor
-    uses_with:
-      traversal_paths: '@c'
-    when:
-      mime_type:
-        $regex: image*
   - name: indexer
     uses: ${{ ENV.INDEXER_NAME }}
-    needs:
```
Comment (azayz) [inline now/app/text_to_video/flow-video-clip-high-performance.yml:58]: maybe this should still need encoder clip ?

File: now/executor/preprocessor/executor.py [hunk 2]
```diff
@@ -59,6 +66,52 @@ def _set_user_input(self, parameters: Dict):
...
+        for doc in flat_docs:
+            text_in_doc = id_to_text[
+                doc.parent_id if self.app.index_query_access_paths == '@c' else doc.id
+            ]
+            doc.tags[TAG_OCR_DETECTOR_TEXT_IN_DOC] = text_in_doc.strip()
+
+    @staticmethod
+    def _save_uri_to_tmp_file(uri, tmpdir) -> str:
+        """Saves URI to a temporary file and returns the path to that file."""
+        req = urllib.request.Request(uri, headers={'User-Agent': 'Mozilla/5.0'})
```
Comment (azayz) [inline now/executor/preprocessor/executor.py:102]: specify timeout parameter so we don't spend much time waiting for the request

File: now/executor/preprocessor/executor.py [hunk 3]
```diff
@@ -59,6 +66,52 @@ def _set_user_input(self, parameters: Dict):
...
                     json.dump(self.user_input.__dict__, fp)
 
+    def _ocr_detect_text(self, docs: DocumentArray):
+        """Iterates over all documents, detects text in images and saves it into the tags of the document."""
+        flat_docs = docs[self.app.index_query_access_paths]
+        # select documents whose mime_type starts with 'image'
+        flat_docs = [
+            doc
+            for doc in flat_docs
+            if doc.mime_type.startswith('image') or doc.modality == 'image'
```
Comment (azayz) [inline now/executor/preprocessor/executor.py:76]: you no longer pass
   when:
      mime_type:
        $regex: image* 
into the flows so the if check is not valid right ?

**Review by azayz — COMMENTED** [review]
LGTM is general, have you done any performance analysis now that we are using paddle paddle ocr ?

File: now/app/text_to_video/flow-video-clip-high-performance.yml (hunk 1 above)
Comment (joschkabraun) [reply now/app/text_to_video/flow-video-clip-high-performance.yml:58]: `needs` isn't needed anymore as we have a linear structure in the flow without branching.

File: now/executor/preprocessor/executor.py (hunk 3 above)
Comment (joschkabraun) [reply now/executor/preprocessor/executor.py:76]: What do you mean with it isn't valid anymore?
We don't do that check in the flow as the preprocessor only applies OCR detection to the relevant documents by the check.

File: now/executor/preprocessor/executor.py (hunk 2 above)
Comment (joschkabraun) [reply now/executor/preprocessor/executor.py:102]: Thank you!

File: now/executor/preprocessor/executor.py [hunk 4]
```diff
@@ -59,6 +66,52 @@ def _set_user_input(self, parameters: Dict):
                 with open(self.user_input_path, 'w') as fp:
                     json.dump(self.user_input.__dict__, fp)
 
+    def _ocr_detect_text(self, docs: DocumentArray):
```
Comment (jupyterjazz) [inline now/executor/preprocessor/executor.py:69]: Have you considered having ocr detection inside preprocess functions? It seems to me that in that case we'll have a better control by knowing which apps and which docs to process.

**Review by jupyterjazz — APPROVED** [review]
LGTM!

File: now/executor/preprocessor/executor.py (hunk 4 above)
Comment (joschkabraun) [reply now/executor/preprocessor/executor.py:69]: I think for now we can decouple it as all apps which involve image modality use CLIP which needs OCR. The OCR detection is also only applied to image documents.

**Pull request merged**